app/queries/keyset_pagination.rb

```diff
@@ -57,17 +56,31 @@ def next_after_key
     end
 
     def is_first_page?
-      @is_first_page ||= ordered_initial_results.empty? ||
-        KeysetPagination.new(
-          client, per_page: 1, before: next_before_key
-        ).empty?
+      if @is_first_page.nil?
```

Comment:
Could this be simplified to 
```    def is_first_page?
      @if_first_page ||=
        # We can assume that a lack of pagination data means we're on the first page
        if previous.nil? || ordered_initial_results.empty?
          true
        else
          where = where_clause(greater_than: order != :asc)
          !client.initial_query.where(where, *next_before_key).exists?
        end
    end```

and same with the is_last_page method?

Reply:
We can't unfortunately as, in the event of a false value, the code we be re-executed each time it's called. I.e if @if_first_page is set to false then `@if_first_page ||=` will be executed each time the method is called. 

Instead this code checks uses a check on whether the ivar is nil which will mean the execution is skipped when it is true or false.

Reply:
Thanks, life would certainly be easier if there was an operator that acted as a "set this if it's not already set" it's frustrating when you want to memoise false and nil values.